src/metabase/driver/generic_sql.clj

```diff
@@ -89,38 +87,56 @@
+  "A map of our currently open connection pools, keyed by DATABASE `:id`."
+  (atom {}))
+
+(defn- create-connection-pool
+  "Create a new C3P0 `ComboPooledDataSource` for connecting to the given DATABASE."
+  [{:keys [id engine details]}]
+  (log/debug (u/format-color 'magenta "Creating new connection pool for database %d ..." id))
+  (let [spec (connection-details->spec (driver/engine->driver engine) details)]
+    (kdb/connection-pool (assoc spec :minimum-pool-size           1
+                                     ;; prevent broken connections closed by dbs by testing them every 3 mins
+                                     :idle-connection-test-period (* 3 60)
+                                     ;; prevent overly large pools by condensing them when connections are idle for 15m+
+                                     :excess-timeout              (* 15 60)))))
+
+(defn- notify-database-updated
+  "We are being informed that a DATABASE has been updated, so lets shut down the connection pool (if it exists) under
+   the assumption that the connection details have changed."
+  [_ {:keys [id]}]
+  (when-let [pool (get @connection-pools id)]
+    (log/debug (u/format-color 'red "Closing connection pool for database %d ..." id))
+    ;; remove the cached reference to the pool so we don't try to use it anymore
+    (reset! connection-pools (dissoc @connection-pools id))
+    ;; now actively shut down the pool so that any open connections are closed
+    (.close ^ComboPooledDataSource (:datasource pool))))
+
+(defn db->pooled-connection-spec
   "Return a JDBC connection spec that includes a cp30 `ComboPooledDataSource`.
-   Theses connection pools are cached so we don't create multiple ones to the same DB.
-   Pools are destroyed after they aren't used for more than 3 hours."
-  (memoize/ttl
-   (fn [spec]
-     (log/debug (u/format-color 'magenta "Creating new connection pool..."))
-     (kdb/connection-pool (assoc spec :minimum-pool-size           1
-                                      ;; prevent broken connections closed by dbs by testing them every 3 mins
-                                      :idle-connection-test-period (* 3 60)
-                                      ;; prevent overly large pools by condensing them when connections are idle for 15m+
-                                      :excess-timeout              (* 15 60))))
-   :ttl/threshold (* 6 60 60 1000)))
+   Theses connection pools are cached so we don't create multiple ones to the same DB."
+  [{:keys [id], :as database}]
+  (if (contains? @connection-pools id)
+    ;; we have an existing pool for this database, so use it
+    (get @connection-pools id)
+    ;; need to create a new pool
+    (let [new-pool (create-connection-pool database)]
+      ;; add the newly created pool to the cache so we maintain a reference to it
+      (reset! connection-pools (assoc @connection-pools id new-pool))
+      ;; just return the newly created pool
+      new-pool)))
```

Comment:
This could be written a little more concisely. Something like this should work:
```clojure
(get @connection-pools id (u/prog1 (create-connection-pool database)
                            (swap! connection-pools assoc id <>)))
```

*  `get` can return a default value so you can create the new pool there
*  `u/prog1` is the business, this is a perfect place to use it
*  prefer `swap!` instead of `reset!`